### PR TITLE
Editor navigation by URLS e2e test

### PR DESCRIPTION
**Pull request by schen2166**
One of the many ways to navigate to the Editor views of the
drawer components is to navogate by URLs according to the
uuid (unique tag at the end of the URL).

These sets of tests test that a user can navigate to the various
editors by changing the URL to one that matchs an existing editor.
These tests navigate from one editor page to another editor page.

This also a subtask for issue <a href="https://github.com/EasyDynamics/oscal-react-library/issues/305"> #305</a>

### CODE REVIEW

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (kylelaker) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:4]: The functions need to use the `base_url` variable and be relative to that

**Comment by kylelaker** [comment]
> navigating to an editor page from another website

Can you elaborate on what you mean by that?

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:4]: There is a way to configure a baseURL in the config file. <a href ="https://docs.cypress.io/api/commands/visit#Prefixes"> Configuring baseURL</a>
Would that be something worth looking into?
It should make code look cleaner because we can just call `cy.visit('/')` to navigate to the base url

Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:4]: > There is a way to configure a baseURL in the config file. 
Nevermind, I just saw it's already defined

**Comment by schen2166** [comment]
> Can you elaborate on what you mean by that?
Suppose you are on the internet browsing on any website. Let's call this *www.example.com* and suddenly for some super rad reason, you want to check on the OSCAL SSP Editor page for updates. so you click on browser url and type in the url to get to the SSP editor (http://localhost:8080/system-security-plan/cff8385f-108e-40a5-8f7a-82f3dc0eaba8) 

Do we want to test for the ability to navigate to these editor pages like this? (assuming that there are special humans who are memorizing these unique `uuid`s )

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (kylelaker) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:23]: These are all pairs of tightly coupled information (the URL and the title). And then we're defining them all as separate values. We also certainly have defined constants like these elsewhere in the code (maybe not these exact ones but I feel pretty confident we've got something similar).

When we have data that will only ever be used together, we should consider grouping it together as an object! In JS this is pretty easy:

```ts
const navigation = [
  {
    title: "NIST Special Publication 800-53 Revision 5: Security and Privacy Controls for Federal Information Systems and Organizations",
    relativeUrl: "/catalog/613fca2d-704a-42e7-8e2b-b206fb92b456",
  },
  {
    title: "MongoDB Component Definition Example",
    relativeUrl: "/component-definition/a7ba800c-a432-44cd-9075-0862cd66da6b"
  },
];
```

and so on.. (I'm not going to type it all out but I think you see the pattern).

And then we can even do other cool things! If you're not familiar with `jest`'s `each` function, you should become familiar because then at least one of the tests could be rewritten as:

```ts
it.each(navigation)("Navigate to $title", ({ title, relativeUrl }) => {
  cy.visit(`${Cypress.env("base_url")}${relativeUrl}`);
  cy.contains(title);
});
```

Which is basically going operate as a `for` loop over the `navigation` list and then execute the test on _each_ of them.

**Review by kylelaker — CHANGES_REQUESTED** [review]

**Comment by kylelaker** [comment]
> Do we want to test for the ability to navigate to these editor pages like this?

I mean, these tests already basically test for that -- Cypress is mimicking a browser the whole time. It's navigating using the full URL for each of these. I don't I see how this is fundamentally different. In any properly functioning browser, I can't think of any reason this would ever be different than doing the same thing from a new tab page. I could see an `<a>` tag potentially polluting with a `Referer` header but not directly typing in the URL bar.

If you can demonstrate a failing test using this, then I'd be interested. But that's a far larger issue than "I was on a Google and just decided to type a full URL".

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:23]: o.O thats really cool I'll try that it looks a lot better that way too.

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:26]: Oopsie, there used to be a before/after each which navigated to the SSP Editor page. I'll remove that describe (line 25) since I no longer think its necessary

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:52]: That;s true its a predefined order an will have the same order every run.

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:39]: I think it would be helpful to have an individual test for each route. Can you place the `it()` inside the loop here? That way you could use the `editor` field inside the test name.

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:38]: The wording of this test is a bit confusing. 

```suggestion
describe("The editor", () => {
  it("loads the correct pages by url", () => {
```

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:41]: Does this work with large files? I would think that the title would not appear instantly. 

#124 adds the `waitForLoad` function that might help with that.

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:1]: I really like the use of `NAVIGATION` here. It makes this a lot more clean.

**Review by tuckerzp — CHANGES_REQUESTED** [review]

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:38]: This is more concise, I like this wording better

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:41]: great idea, I added a `waitForLoad` function call

Comment (tuckerzp) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:41]: You do not need to add it, as it will be added in that other pr. Just use it when that pr is merged in.

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:39]: Is it not as simple as moving that `it` to inside the `each` body?

Comment (schen2166) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:39]: it appears to be a little more complex than just moving the it inside the `each`

File: end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js
Comment (tuckerzp) [inline end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:44]: You can map each `Navigation` item to a test.
```
  Navigation.map((oscalItem) => {} ...);
```

Comment (kylelaker) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:44]: Please don't do this either. Look into `test.each`/`it.each` and try to get that to work. Let's not `map` our tests unless we need to.

If it doesn't work, as I've asked, please be descriptive about the error encountered instead of just trying something else. `each` or a Cypress tool that mimics it is the right thing to do here.

Comment (tuckerzp) [reply end-to-end-tests/cypress/e2e/OSCAL/url_navigation.cy.js:44]: https://jestjs.io/docs/api#testeachtablename-fn-timeout

Jest provides a `test.each` function that is to be used when "you keep duplicating the same test with different data". That most accurately describes the issue we are trying to fix here. I probably should have better looked at the docs before recommending the `map`. 

I think in general reinventing the wheel is something to avoid. Using functionality provided by `Jest` or `Cypress` should be the first thing you attempt before trying other options. Since this seems like a fairly common thing you would want to test, it is best to see if there is something that already achieves what we want.

**Comment by schen2166** [comment]
Here are the issues I am encountering with using `it.each`
Following the `it.each(table)(name, fn)` syntax I get an error saying the function doesn't exist


Function seems to exist, however cannot find online references to resolve this error. It is also possible that the error there occurs because element is not a defined function. However changing the `element` variable to one that is defined results in error under (wider image).


`test.each` Does not seem to be a function I can use to test navigation
Test.each attempt
Test.each does not seem to be a function

This would be the function call to iterate through an array found (Cypress' <a href ="https://docs.cypress.io/api/commands/each#Syntax"> `.each` function reference link</a>) which also wouldn't work because the `it` statements cannot be nested inside an each statement.

**Comment by mikeisen1** [comment]
@schen2166, we're now getting a `cy.waitForLoad is not a function`. I think for now, while waiting for #124 to be merged, we should just use `cy.wait`.

**Comment by schen2166** [comment]
> @schen2166, we're now getting a `cy.waitForLoad is not a function` error. I think for now, while waiting for #124 to be merged, 

Yes, we will need to wait for Zach to merge in his task. I've coordinated it with him. He said that we should wait to merge this issue in till after he merges his feature in. We should leave the code as is for now.

**Review by kylelaker — APPROVED** [review]
Thanks for going back and forth on the `it.each` stuff! With the additional context about what the error was, it now makes sense what's happening and we can move forward with `.forEach`. This is a really nice improvement.

**Pull request merged**
